### ISSUE TITLE
BigEarthNet preprocessing producing very low RGB values for S2 images

### ISSUE THREAD

**Issue by mmendiet** [microsoft/torchgeo#549]
I am just starting to play with the BigEarthNet dataset and preprocessing code. However, after loading the dataset and attempting to visualize some examples, I found that they all come out almost black.
Visualization of sample retrieved by val_dataset.__getitem__(0)
![s2_ben0](https://user-images.githubusercontent.com/26916118/171771425-ae02b8b9-61f4-4f2b-a671-a44ad22bcde2.png)
Visualization of sample retrieved by val_dataset.__getitem__(100)
![s2_ben100](https://user-images.githubusercontent.com/26916118/171771440-b261b60e-9a18-42b7-84bd-d70506bd5f44.png)
Visualization of sample retrieved by val_dataset.__getitem__(1000)
![s2_ben1000](https://user-images.githubusercontent.com/26916118/171771449-0e1024f3-1ae5-4f14-ad19-0a3f500935be.png)

The preprocessing code I am referring to is: https://github.com/microsoft/torchgeo/blob/d06b103f81edec4f4e0d13ccd621d318364679a2/torchgeo/datamodules/bigearthnet.py#L110
```
 band_mins = torch.tensor(
        [-48.0, -42.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
    )
 band_maxs = torch.tensor(
        [
            6.0,
            16.0,
            9859.0,
            12872.0,
            13163.0,
            14445.0,
            12477.0,
            12563.0,
            12289.0,
            15596.0,
            12183.0,
            9458.0,
            5897.0,
            5544.0,
        ]
    )
```
and
```
        if bands == "all":
            self.mins = self.band_mins[:, None, None]
            self.maxs = self.band_maxs[:, None, None]
        elif bands == "s1":
            self.mins = self.band_mins[:2, None, None]
            self.maxs = self.band_maxs[:2, None, None]
        else:
            self.mins = self.band_mins[2:, None, None]
            self.maxs = self.band_maxs[2:, None, None]

    def preprocess(self, sample: Dict[str, Any]) -> Dict[str, Any]:
        """Transform a single sample from the Dataset."""
        sample["image"] = sample["image"].float()
        sample["image"] = (sample["image"] - self.mins) / (self.maxs - self.mins)
        sample["image"] = torch.clip(sample["image"], min=0.0, max=1.0)
        return sample
```
After these preprocessing operations, the RGB pixel values are very low. For example, the RGB channels for the sample retrieved by val_dataset.__getitem__(0):
```
array([[[0.05843257, 0.06359216, 0.04700489],
        [0.06487123, 0.07969914, 0.06580685],
        [0.08917198, 0.09314694, 0.08693963],
        ...,
        [0.05642481, 0.07118979, 0.03729314],
        [0.07117142, 0.07772375, 0.04599487],
        [0.06293271, 0.07445677, 0.0417217 ]],

       [[0.03295486, 0.05067619, 0.02991221],
        [0.05732484, 0.067315  , 0.05679435],
        [0.0744946 , 0.07924327, 0.06479683],
        ...,
        [0.04465522, 0.0644279 , 0.03247611],
        [0.04707837, 0.06047713, 0.03270919],
        [0.03385489, 0.04361039, 0.02618289]],

       [[0.02312379, 0.0525756 , 0.02408515],
        [0.02776239, 0.0499924 , 0.02719291],
        [0.04105511, 0.05888163, 0.03729314],
        ...,
        [0.03025478, 0.05310743, 0.0261052 ],
        [0.02222376, 0.04186294, 0.02043353],
        [0.01765439, 0.03434129, 0.01600497]],

       ...,

       [[0.06971753, 0.07141772, 0.04358636],
        [0.0690252 , 0.0692144 , 0.04413021],
        [0.06286347, 0.0670111 , 0.04312019],
        ...,
        [0.02049294, 0.04512992, 0.02004506],
        [0.03918582, 0.05880565, 0.03208764],
        [0.07047909, 0.0759763 , 0.04747106]],

       [[0.07047909, 0.07324115, 0.04343097],
        [0.06715591, 0.06951831, 0.03946857],
        [0.06424813, 0.06989819, 0.0429648 ],
        ...,
        [0.02720853, 0.05067619, 0.02284205],
        [0.05621712, 0.06503571, 0.04226556],
        [0.08937968, 0.08608114, 0.0574159 ]],

       [[0.07712545, 0.07445677, 0.0456064 ],
        [0.0702714 , 0.07255736, 0.04381944],
        [0.07089449, 0.06997417, 0.04482946],
        ...,
        [0.03039324, 0.05508281, 0.02664906],
        [0.05130158, 0.06214861, 0.03636081],
        [0.08709498, 0.0836499 , 0.05539585]]], dtype=float32)
```
Note, I already removed the scaling line here when plotting the normalized data. https://github.com/microsoft/torchgeo/blob/d06b103f81edec4f4e0d13ccd621d318364679a2/torchgeo/datasets/bigearthnet.py#L549
The values and plots shown above are just the data tensor after the preprocessing function.

On the other hand, I tried some other normalization values from the SeCo code base:
https://github.com/ElementAI/seasonal-contrast/blob/5395c027922569f5c5b1785ad1ccddd839749c36/datasets/bigearthnet_dataset.py#L13
```
mean = torch.tensor([
        340.76769064,
        429.9430203,
        614.21682446,
        590.23569706,
        950.68368468,
        1792.46290469,
        2075.46795189,
        2218.94553375,
        2266.46036911,
        2246.0605464,
        1594.42694882,
        1009.32729131
    ])
    std = torch.tensor([
        554.81258967,
        572.41639287,
        582.87945694,
        675.88746967,
        729.89827633,
        1096.01480586,
        1273.45393088,
        1365.45589904,
        1356.13789355,
        1302.3292881,
        1079.19066363,
        818.86747235
    ])
```
and modified the preprocessing function:
```
    def preprocess(self, sample: Dict[str, Any]) -> Dict[str, Any]:
        """Transform a single sample from the Dataset."""
        self.mins = (self.mean - 2 * self.std)[:, None, None] ## modified 
        self.maxs = (self.mean + 2 * self.std)[:, None, None] ## modified
        sample["image"] = sample["image"].float()
        sample["image"] = (sample["image"] - self.mins) / (self.maxs - self.mins)
        sample["image"] = torch.clip(sample["image"], min=0.0, max=1.0)
        return sample
```
Here are the same images as before, but now with the SeCo preprocessing values:
![s2_seco0](https://user-images.githubusercontent.com/26916118/171771615-ea338ced-129c-40d0-a1b0-01181381ff35.png)
![s2_seco100](https://user-images.githubusercontent.com/26916118/171771631-fae01412-83bc-4e99-bfb3-8fe8376f3f8a.png)
![s2_seco1000](https://user-images.githubusercontent.com/26916118/171771636-2c724ebf-c6a4-458d-b970-9a7a5990c790.png)

And it seems to visualize well. Is there a particular reason for the use of the provided band_mins and band_maxs values in this repo for BigEarthNet? If I am missing something obvious, please let me know; I am new to the geospatial domain. Thanks!

**Comment by calebrob6** [comment]
Hey @mmendiet, thanks for opening an issue!

I'm seeing that plotting is working as expected (although the class names look pretty dumb...):
![image](https://user-images.githubusercontent.com/1088078/173770650-0074cab4-65ee-4906-bdd6-f76e4cbd2060.png)


I think you might be trying to visualize the imagery after it goes through the preprocessing step that happens for model training / inference. A standard preprocessing step is to perform channel-wise standardization, i.e. subtract the dataset mean per channel from each image, then divide by the dataset standard deviation per channel. This centers values around 0, and keeps the values small (more/less) -- which deep learning models like (more/less).

See https://github.com/microsoft/torchgeo/blob/main/torchgeo/datasets/bigearthnet.py#L526 for the plot function that takes the raw values from the Sentinel 2 imagery and visualizes them as RGB.

**Comment by calebrob6** [comment]
I'm going to go ahead and close this as there isn't an issue with the BigEarthNet plotting functionality, but feel free to follow up with questions / comments!
